data_structures/binary_search_tree.cpp

```diff
@@ -1,138 +1,246 @@
 /**
  * \file
  * \brief A simple tree implementation using structured nodes
- *
- * \todo update code to use C++ STL library features and OO structure
- * \warning This program is a poor implementation - C style - and does not
- * utilize any of the C++ STL features.
  */
+#include <algorithm>  // test code
+#include <cassert>
 #include <iostream>
+#include <memory>
+#include <queue>
+#include <vector>
 
-struct node {
-    int val;
-    node *left;
-    node *right;
-};
+/**
+ * @namespace data_structure
+ * @brief Implementations of various data structures
+ */
+namespace data_structure {
+/**
+ * @namespace data_structure::BST
+ * @brief Implementation of Binary Search Tree data structure
+ */
+namespace BST {
 
-struct Queue {
-    node *t[100];
-    int front;
-    int rear;
+/** An struct node type.
+ *  Define a element for tree insert/remove operation
+ */
+struct node {
+    int val{};                   /**< int value of a node struct*/
+    std::shared_ptr<node> left;  /**< left subtree pointer */
+    std::shared_ptr<node> right; /**< right subtree pointer */
+    //~node() { std::cout << "node: destructor, val = " << val << '\n'; }
 };
 
-Queue queue;
-
-void enqueue(node *n) { queue.t[queue.rear++] = n; }
-
-node *dequeue() { return (queue.t[queue.front++]); }
-
-void Insert(node *n, int x) {
+/** insert a node to tree
+ * with greater value to right subtree, smaller value node to left subtree
+ * \param[in] n root node of a tree
+ * \param[in] x a node with value to be insert
+ */
+void Insert(node* n, int x) {
+    assert(n != NULL);
```

Comment:
what happens if the user inserts the very first root element? this Insert function will break in that case? you are not handling the root insertition case, instead you assumed to insert the root value manually as mentioned in the test function
```
    root->val = 4;
    root->left = nullptr;
    root->right = nullptr;
```
i suggest you to handle this too in the insert function because basically we are performing the very same operation that is we are insterting a value let it be anywhere in the tree either the root or the child everything should be properly handled by the same insert function

Reply:
@ayaankhan98 Thanks for you suggestion. I have got your means: 
you want to design a function that is very common to use.
But , since here we are talking about a special tree (binary search tree).

Reply:
https://github.com/ayaankhan98/DSA_competative/blob/b1d200d8a9614141fa1305f1e41c605526e91336/Data_Structures/Trees/BST.cpp#L25

have a look at this code that which i had written a year ago, i hope this makes you clear what i wish to say.

```
if (root == nullptr) {
        Node* new_node = new Node;
        new_node->data = data;
        new_node->left = nullptr;
        new_node->right = nullptr;
        root = new_node;
    }
```
if the root is null then we can do something like this. @liushubin-gitHub  And i think this would be better instead of inserting the root node manually like
```
    root->val = 4;
    root->left = nullptr;
    root->right = nullptr;
```

Reply:
Got it.Thanks a lot!

Reply:
Nice look at that link i had implemented some extra methods on BST, see if you can implement that in this PR.
@liushubin-gitHub  if you face any problem let me know.